### PR TITLE
Using STL vector in locate_dofs_topological

### LINKED ISSUE #1348: locate_dofs_topological still using Eigen vector
A recent API change introduced the use of `std::vector` in place of Eigen arrays. `locate_dofs_topological` was left out and still takes an Eigen array as a parameter. This makes it impossible, for example, to pass directly the value returned by `MeshTags::find`, which is a natural concatenation.

### PR DESCRIPTION
**Pull request by massimiliano-leoni**
Closes #1348 

Changing `locate_dofs_topological` to take an `std::vector` as an argument.

### CODE REVIEW

File: cpp/dolfinx/fem/DirichletBC.cpp [hunk 1]
```diff
@@ -409,7 +408,7 @@ fem::locate_dofs_topological(const fem::FunctionSpace& V, const int dim,
   const int num_entity_closure_dofs
       = dofmap->element_dof_layout->num_entity_closure_dofs(dim);
   std::vector<std::int32_t> dofs;
-  for (Eigen::Index i = 0; i < entities.rows(); ++i)
+  for (size_t i = 0; i < entities.size(); ++i)
```
Comment (IgorBaratta) [inline cpp/dolfinx/fem/DirichletBC.cpp:411]: I think the C ++ version `std :: size_t` is preferable.

Comment (massimiliano-leoni) [reply cpp/dolfinx/fem/DirichletBC.cpp:411]: Of course, that was unintentional :D
Thanks!

File: python/dolfinx/wrappers/fem.cpp [hunk 2]
```diff
@@ -457,12 +457,14 @@ void fem(py::module& m)
...
-         const int dim, const Eigen::Ref<const Eigen::ArrayXi>& entities,
+         const int dim, const py::array_t<std::int32_t>& entities,
          bool remote) -> std::array<py::array, 2> {
         if (V.size() != 2)
           throw std::runtime_error("Expected two function spaces.");
+        auto sp
+            = tcb::span<const std::int32_t>(entities.data(), entities.size());
         std::array<std::vector<std::int32_t>, 2> dofs
-            = dolfinx::fem::locate_dofs_topological({V[0], V[1]}, dim, entities,
+            = dolfinx::fem::locate_dofs_topological({V[0], V[1]}, dim, sp,
```
Comment (garth-wells) [inline python/dolfinx/wrappers/fem.cpp:467]: Replace `sp` with `tcb::span(entities.data(), entities.size())`

File: python/dolfinx/wrappers/fem.cpp [hunk 3]
```diff
@@ -471,9 +473,11 @@ void fem(py::module& m)
   m.def(
       "locate_dofs_topological",
       [](const dolfinx::fem::FunctionSpace& V, const int dim,
-         const Eigen::Ref<const Eigen::ArrayXi>& entities, bool remote) {
+         const py::array_t<std::int32_t>& entities, bool remote) {
+        auto sp
+            = tcb::span<const std::int32_t>(entities.data(), entities.size());
         return as_pyarray(
-            dolfinx::fem::locate_dofs_topological(V, dim, entities, remote));
+            dolfinx::fem::locate_dofs_topological(V, dim, sp, remote));
```
Comment (garth-wells) [inline python/dolfinx/wrappers/fem.cpp:480]: See comment above. Uses modern C++ type inference.

**Review by garth-wells — CHANGES_REQUESTED** [review]
Looks good just some small style comments.

File: python/dolfinx/wrappers/fem.cpp (hunk 2 above)
Comment (massimiliano-leoni) [reply python/dolfinx/wrappers/fem.cpp:467]: That's how I first tried it, but it complained that

```
fem.cpp:467:15: error: no matching function for call to 'locate_dofs_topological'
DirichletBC.h:53:42: note: candidate function not viable: expects an l-value for 3rd argument
```

File: cpp/dolfinx/fem/DirichletBC.h [hunk 4]
```diff
@@ -78,7 +77,7 @@ std::array<std::vector<std::int32_t>, 2> locate_dofs_topological(
 /// with V.
 std::vector<std::int32_t>
 locate_dofs_topological(const fem::FunctionSpace& V, const int dim,
-                        const Eigen::Ref<const Eigen::ArrayXi>& entities,
+                        tcb::span<const std::int32_t>& entities,
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.h:80]: Need to add const here.

File: python/dolfinx/wrappers/fem.cpp (hunk 2 above)
Comment (garth-wells) [reply python/dolfinx/wrappers/fem.cpp:467]: There is a missing `const` - I just posted a comment.

File: python/dolfinx/wrappers/fem.cpp [hunk 5]
```diff
@@ -457,23 +457,26 @@ void fem(py::module& m)
...
   m.def(
       "locate_dofs_topological",
       [](const dolfinx::fem::FunctionSpace& V, const int dim,
-         const Eigen::Ref<const Eigen::ArrayXi>& entities, bool remote) {
-        return as_pyarray(
-            dolfinx::fem::locate_dofs_topological(V, dim, entities, remote));
+         const py::array_t<std::int32_t>& entities, bool remote) {
+        return as_pyarray(dolfinx::fem::locate_dofs_topological(
+            V, dim,
+            tcb::span<const std::int32_t>(entities.data(), entities.size()),
```
Comment (garth-wells) [inline python/dolfinx/wrappers/fem.cpp:478]: Use `tcb::span(entities.data(), entities.size())`

Comment (massimiliano-leoni) [reply python/dolfinx/wrappers/fem.cpp]: Was already on the way!

**Review by garth-wells — APPROVED** [review]

**Pull request merged**